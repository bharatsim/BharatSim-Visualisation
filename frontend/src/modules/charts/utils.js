import { chartStyleConfig, getColor } from './chartStyleConfig';

function trasformDataForChart(csvData, xColumn, yColumns) {
  return {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
}

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export { trasformDataForChart, getYaxisNames };
