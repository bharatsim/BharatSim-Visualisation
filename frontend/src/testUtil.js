// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent, within } from '@testing-library/react';

export function selectDropDownOption(container, dropDownId, optionId) {
  const dropDown = container.getByTestId(dropDownId);
  fireEvent.mouseDown(within(dropDown).getByRole('button'));
  const options = within(
    within(document.getElementById(`menu-${dropDownId}`)).getByRole('listbox'),
  );
  fireEvent.click(options.getByTestId(`${dropDownId}-${optionId}`));
}
