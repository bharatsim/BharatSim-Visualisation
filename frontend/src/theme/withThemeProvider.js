import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
