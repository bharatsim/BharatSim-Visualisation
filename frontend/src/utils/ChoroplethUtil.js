import L from 'leaflet';
import chroma from 'chroma-js';

function highlightOnClick(e) {
  const layer = e.target;

  layer.setStyle({
    weight: 2,
    dashArray: '',
    opacity: 1,
    fillOpacity: 0.7,
  });

  if (!L.Browser.ie && !L.Browser.opera && !L.Browser.edge) {
    layer.bringToFront();
  }
}

function resetHighlight(geoJson) {
  return (e) => geoJson.current.resetStyle(e.target);
}

function zoomToFeature(map) {
  return (e) => map.fitBounds(e.target.getBounds());
}

function geoJSONStyle(idDataMap, mapLayerIdName, maxOfMeasure, scale) {
  const colors = Object.keys(scale)
    .sort((v1, v2) => Number(v1) - Number(v2))
    .map((key) => scale[key]);
  const getColor = chroma.scale(colors);
  return (feature) => {
    const gradientPercentage = idDataMap[feature.properties[mapLayerIdName]] / maxOfMeasure;
    return {
      color: '#1f2021',
      weight: 1,
      fillOpacity: 0.5,
      opacity: 0.7,
      fillColor: getColor(gradientPercentage),
    };
  };
}

export { highlightOnClick, resetHighlight, geoJSONStyle, zoomToFeature };
