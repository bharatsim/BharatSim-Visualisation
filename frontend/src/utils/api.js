import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  addNewDashboard: async ({ name, projectId }) => {
    return uploadData({
      url: serviceURL.INSERT_DASHBOARD,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets: [], layout: [], name, count: 0, projectId },
      }),
      isCustomErrorHandler: true,
    });
  },

  getAllDashBoardByProjectId: async (projectId) =>
    fetchData({
      url: serviceURL.DASHBOARD_URL,
      query: { projectId, columns: ['name', '_id'] },
    }),

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema, dashboardId }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
        { name: 'dashboardId', value: dashboardId },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async (dashboardId) =>
    fetchData({
      url: serviceURL.DATA_SOURCES,
      query: { dashboardId },
    }),

  getData: async (datasource, columns) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query: { columns } }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      isCustomErrorHandler: true,
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
