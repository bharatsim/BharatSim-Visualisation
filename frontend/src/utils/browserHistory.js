// eslint-disable-next-line import/no-mutable-exports
let BrowserHistory;

function initHistory(history) {
  BrowserHistory = history;
}

function navigateToPreviousPage() {
  BrowserHistory.goBack();
}

function navigateToHome() {
  BrowserHistory.push('/');
}

function reloadPage() {
  // eslint-disable-next-line no-restricted-globals
  location.reload();
}

export { initHistory, navigateToPreviousPage, reloadPage, BrowserHistory, navigateToHome };
