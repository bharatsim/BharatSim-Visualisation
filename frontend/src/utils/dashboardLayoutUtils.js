export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 6) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 6,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
