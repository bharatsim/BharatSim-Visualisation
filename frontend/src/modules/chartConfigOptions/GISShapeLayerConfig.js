import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import { Box, Typography } from '@material-ui/core';
import { projectLayoutContext } from '../../contexts/projectLayoutContext';
import useLoader from '../../hook/useLoader';
import { api } from '../../utils/api';
import { convertObjectArrayToOptionStructure } from '../../utils/helper';
import Dropdown from '../../uiComponent/Dropdown';
import LoaderOrError from '../loaderOrError/LoaderOrError';
import NoDataSetPresentMessage from '../configureDataset/NoDatatSetPresentMessage';

const GISShapeLayerFileTypes = ['geojson', 'topojson', 'json'];

function GISShapeLayerConfig({ handleConfigChange, configKey, error, value }) {
  const { selectedDashboardMetadata, projectMetadata } = useContext(projectLayoutContext);
  const { _id: selectedDashboardId } = selectedDashboardMetadata;
  const {
    startLoader,
    stopLoaderAfterSuccess,
    stopLoaderAfterError,
    loadingState,
    message,
  } = useLoader();
  const [fetchedDatasources, setFetchedDatasources] = useState();

  useEffect(() => {
    fetchDatasources();
  }, []);

  async function fetchDatasources() {
    startLoader();
    api
      .getDatasources(selectedDashboardId)
      .then((resData) => {
        stopLoaderAfterSuccess();
        setFetchedDatasources(resData);
      })
      .catch(() => {
        stopLoaderAfterError('Unable to fetch data sources');
      });
  }

  const dataSources = fetchedDatasources
    ? fetchedDatasources.dataSources.filter((dataSource) =>
        GISShapeLayerFileTypes.includes(dataSource.fileType),
      )
    : [];

  const isDataSourcePresent = dataSources.length > 0;

  const onErrorAction = {
    name: 'Retry',
    onClick: fetchDatasources,
  };

  function handleChange(selectedValue) {
    handleConfigChange(configKey, selectedValue);
  }

  return (
    <LoaderOrError
      loadingState={loadingState}
      message={message}
      errorAction={onErrorAction}
      fullWidth
    >
      <Box px={2}>
        {isDataSourcePresent ? (
          <>
            <Box mb={2}>
              <Typography variant="subtitle2">GIS Shape Layer</Typography>
            </Box>
            <Dropdown
              options={convertObjectArrayToOptionStructure(dataSources, 'name', '_id')}
              onChange={handleChange}
              id="dropdown-gis-shape-layer"
              label="select GIS shape Layer"
              error={error}
              value={value}
            />
          </>
        ) : (
          <Box>
            <NoDataSetPresentMessage
              projectMetadataId={projectMetadata.id}
              message="Before we can create any GIS visualization, weâ€˜ll need some GIS layer data."
            />
          </Box>
        )}
      </Box>
    </LoaderOrError>
  );
}

GISShapeLayerConfig.defaultProps = {
  error: '',
  value: '',
};

GISShapeLayerConfig.propTypes = {
  handleConfigChange: PropTypes.func.isRequired,
  error: PropTypes.string,
  value: PropTypes.string,
  configKey: PropTypes.string.isRequired,
};

export default GISShapeLayerConfig;
