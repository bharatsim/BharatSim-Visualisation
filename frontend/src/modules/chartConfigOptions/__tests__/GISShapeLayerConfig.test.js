import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';

import { selectDropDownOption, withProjectLayout, withRouter } from '../../../testUtil';
import withThemeProvider from '../../../theme/withThemeProvider';
import GISShapeLayerConfig from '../GISShapeLayerConfig';
import { api } from '../../../utils/api';

jest.mock('../../../utils/api', () => ({
  api: {
    getDatasources: jest.fn().mockResolvedValue({
      dataSources: [
        { name: 'datasource1', _id: 'id1', fileType: 'csv' },
        { name: 'datasource2', _id: 'id2', fileType: 'geojson' },
        { name: 'datasource3', _id: 'id3', fileType: 'json' },
      ],
    }),
  },
}));

describe('<GISShapeLayerConfig />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  const GISShapeLayerConfigWithProvides = withThemeProvider(
    withRouter(withProjectLayout(GISShapeLayerConfig)),
  );
  it('should match snapshot for GIS Shape Layer selector', async () => {
    const { container, getByText } = render(
      <GISShapeLayerConfigWithProvides
        handleConfigChange={jest.fn()}
        value=""
        error=""
        configKey="gisShapeLayer"
      />,
    );

    await waitFor(() => getByText('GIS Shape Layer'));

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for GIS Shape Layer selector in editMode', async () => {

    const { container, getByText } = render(
      <GISShapeLayerConfigWithProvides
        handleConfigChange={jest.fn()}
        value="id2"
        error=""
        isEditMode
        configKey="gisShapeLayer"
      />,
    );

    await waitFor(() => getByText('GIS Shape Layer'));


    expect(container).toMatchSnapshot();
  });

  it('should show no gis shape file present message if api have no gis files', async () => {
    api.getDatasources.mockResolvedValueOnce({
      dataSources: [{ name: 'datasource1', _id: 'id1', fileType: 'csv' }],
    });

    const { findByText, getByText } = render(
      <GISShapeLayerConfigWithProvides
        handleConfigChange={jest.fn()}
        value=""
        error=""
        configKey="gisShapeLayer"
      />,
    );

    await findByText('Before we can create any GIS visualization, we‘ll need some GIS layer data.');
    expect(
      getByText('Before we can create any GIS visualization, we‘ll need some GIS layer data.'),
    ).toBeInTheDocument();
  });

  it('should handle  GIS Shape Layer change on click of different GIS Shape Layer name', async () => {
    const handleConfigChange = jest.fn();
    const renderedComponent = render(
      <GISShapeLayerConfigWithProvides
        configKey="gisShapeLayer"
        handleConfigChange={handleConfigChange}
        value=""
        error=""
      />,
    );

    await waitFor(() => renderedComponent.getByText('GIS Shape Layer'));

    selectDropDownOption(renderedComponent, 'dropdown-gis-shape-layer', 'datasource2');
    expect(handleConfigChange).toHaveBeenCalledWith('gisShapeLayer', 'id2');
  });

  it('should show loader while fetching data', async () => {
    const handleConfigChange = jest.fn();
    const renderedComponent = render(
      <GISShapeLayerConfigWithProvides
        configKey="gisShapeLayer"
        handleConfigChange={handleConfigChange}
        value=""
        error=""
      />,
    );
    const loaderComponent = document.getElementsByTagName('svg');

    expect(loaderComponent).not.toBeNull();

    await renderedComponent.findByText('select GIS shape Layer');
  });

  it('should show error if error occur while fetching data', async () => {
    api.getDatasources.mockRejectedValueOnce('error');
    const { findByText, getByText } = render(
      <GISShapeLayerConfigWithProvides
        configKey="gisShapeLayer"
        handleConfigChange={jest.fn()}
        value=""
        error=""
      />,
    );

    await findByText('Unable to fetch data sources');

    expect(getByText('Unable to fetch data sources')).toBeInTheDocument();
  });

  it('should refetch data on click on retry button present on error banner', async () => {
    api.getDatasources.mockRejectedValueOnce('error');
    const { findByText, getByText } = render(
      <GISShapeLayerConfigWithProvides
        configKey="gisShapeLayer"
        handleConfigChange={jest.fn()}
        value=""
        error=""
      />,
    );

    await findByText('Unable to fetch data sources');
    const retryButton = getByText('Retry').closest('button');
    fireEvent.click(retryButton);

    await findByText('select GIS shape Layer');

    expect(api.getDatasources).toHaveBeenCalled();
  });
});
