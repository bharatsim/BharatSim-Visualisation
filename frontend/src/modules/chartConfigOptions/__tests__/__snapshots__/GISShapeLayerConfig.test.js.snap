// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GISShapeLayerConfig /> should match snapshot for GIS Shape Layer selector 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-3"
  >
    <div
      class="MuiBox-root MuiBox-root-4"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        GIS Shape Layer
      </h6>
    </div>
    <div
      class="MuiFormControl-root makeStyles-formControl-12"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
        data-shrink="false"
        id="dropdown-label"
      >
        select GIS shape Layer
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
        data-testid="dropdown-gis-shape-layer"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="dropdown-label dropdown-gis-shape-layer"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
          id="dropdown-gis-shape-layer"
          role="button"
          tabindex="0"
        >
          <span>
            â€‹
          </span>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<GISShapeLayerConfig /> should match snapshot for GIS Shape Layer selector in editMode 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-19"
  >
    <div
      class="MuiBox-root MuiBox-root-20"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        GIS Shape Layer
      </h6>
    </div>
    <div
      class="MuiFormControl-root makeStyles-formControl-28"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
        data-shrink="true"
        id="dropdown-label"
      >
        select GIS shape Layer
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline Mui-disabled Mui-disabled MuiInputBase-formControl"
        data-testid="dropdown-gis-shape-layer"
      >
        <div
          aria-disabled="true"
          aria-haspopup="listbox"
          aria-labelledby="dropdown-label dropdown-gis-shape-layer"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input Mui-disabled Mui-disabled Mui-disabled"
          id="dropdown-gis-shape-layer"
          role="button"
        >
          datasource2
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value="id2"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled Mui-disabled"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
